File: 2019/Haskell/1/main.hs
#!/usr/bin/env stack
-- stack --system-ghc --resolver lts-14.16 script

import Control.Applicative ((<**>))
import qualified Data.Text    as Text
import qualified Data.Text.IO as Text

main :: IO ()
main = Text.readFile "input.txt" 
    <**> return (((map $ read . Text.unpack) . Text.lines) :: Text.Text -> [Integer]) 
    >>= \ints -> do
        let totalMassBeforeAccounting = foldl (\acc mass -> fuelForModuleWithMass mass + acc) 0 ints
        putStrLn $ "Total fuel before accounting for fuel mass: " ++ show totalMassBeforeAccounting
        let totalMassAfterAccounting = foldl (\acc mass -> fuelForModuleWithMassAccountingForFuelMass mass + acc) 0 ints
        putStrLn $ "Total fuel after accounting for fuel mass: " ++ show totalMassAfterAccounting

fuelForModuleWithMass :: Integer -> Integer
fuelForModuleWithMass mass = (mass `div` 3) - 2

fuelForModuleWithMassAccountingForFuelMass :: Integer -> Integer
fuelForModuleWithMassAccountingForFuelMass mass
                                            | mass < 0 = 0
                                            | otherwise = let fuelMass = fuelForModuleWithMass mass in
                                                            fuelMass + fuelForModuleWithMassAccountingForFuelMass fuelMass   

File: 2019/Elixir/1/main.exs
require File
require IO

defmodule Main do

  # Part 1
  def totalFuel() do
    case File.read "input.txt" do
      {:ok, contents} -> 
    modulesMasses = contents 
    |> String.split("\n", trim: true)
        |> Enum.map(&String.to_integer/1)
    
    modulesMasses 
        |> Enum.reduce(0, fn(mass, acc) -> acc + fuelForModuleWithMass mass end)
        |> fn total -> IO.puts "Total fuel before fuel mass addition: #{total}" end.()
        
    modulesMasses 
        |> Enum.reduce(0, fn(mass, acc) -> acc + fuelForModuleWithMassAccountingForFuelMass mass end)
        |> fn total -> IO.puts "Total fuel after fuel mass addition: #{total}" end.()
      {:error, reason} ->
    IO.puts("Cannot open input.txt: #{reason}")
    end
  end
    
  def fuelForModuleWithMass(mass) do
    div(mass, 3) - 2
  end
  
   # Part 2
   def fuelForModuleWithMassAccountingForFuelMass(mass) do
    massForMudule = fuelForModuleWithMass mass
     if massForMudule > 0 do
      massForMudule + fuelForModuleWithMassAccountingForFuelMass massForMudule
     else
       0
     end
   end  

end
